packages/browser-agent-core/features/ajax/instrument/index.js

```diff
@@ -0,0 +1,391 @@
+  }
+
+  // Listen for those load listeners to be called.
+  function onFnStart(args, xhr, methodName) {
+    if (xhr instanceof origXHR) {
+      if (methodName === 'onload') this.onload = true
+      if ((args[0] && args[0].type) === 'load' || this.onload) this.xhrCbStart = now()
+    }
+  }
+
+  function onFnEnd(args, xhr) {
+    if (this.xhrCbStart) ee.emit('xhr-cb-time', [now() - this.xhrCbStart, this.onload, xhr], xhr)
+  }
+
+  // this event only handles DT
+  function onFetchBeforeStart(args) {
+    var opts = args[1] || {}
+    var url
+    // argument is USVString
+    if (typeof args[0] === 'string') {
+      url = args[0]
+      // argument is Request object
+    } else if (args[0] && args[0].url) {
+      url = args[0].url
+      // argument is URL object
+    } else if (window.URL && args[0] && args[0] instanceof URL) {
+      url = args[0].href
+    }
+
+    if (url) {
+      this.parsedOrigin = parseUrl(url)
+      this.sameOrigin = this.parsedOrigin.sameOrigin
+    }
+
+    var payload = dt.generateTracePayload(this.parsedOrigin)
+    if (!payload || (!payload.newrelicHeader && !payload.traceContextParentHeader)) {
+      return
+    }
+
+    if (typeof args[0] === 'string' || (window.URL && args[0] && args[0] instanceof URL)) {
+      var clone = {}
+
+      for (var key in opts) {
+        clone[key] = opts[key]
+      }
+
+      clone.headers = new Headers(opts.headers || {})
+      if (addHeaders(clone.headers, payload)) {
+        this.dt = payload
+      }
+
+      if (args.length > 1) {
+        args[1] = clone
+      } else {
+        args.push(clone)
+      }
+    } else if (args[0] && args[0].headers) {
+      if (addHeaders(args[0].headers, payload)) {
+        this.dt = payload
+      }
+    }
+
+    function addHeaders(headersObj, payload) {
+      var added = false
+      if (payload.newrelicHeader) {
+        headersObj.set('newrelic', payload.newrelicHeader)
+        added = true
+      }
+      if (payload.traceContextParentHeader) {
+        headersObj.set('traceparent', payload.traceContextParentHeader)
+        if (payload.traceContextStateHeader) {
+          headersObj.set('tracestate', payload.traceContextStateHeader)
+        }
+        added = true
+      }
+      return added
+    }
+  }
+
+  function onFetchStart(fetchArguments, dtPayload) {
+    this.params = {}
+    this.metrics = {}
+    this.startTime = now()
+    this.dt = dtPayload
+
+    if (fetchArguments.length >= 1) this.target = fetchArguments[0]
+    if (fetchArguments.length >= 2) this.opts = fetchArguments[1]
+
+    var opts = this.opts || {}
+    var target = this.target
+
+    var url
+    if (typeof target === 'string') {
+      url = target
+    } else if (typeof target === 'object' && target instanceof origRequest) {
+      url = target.url
+    } else if (window.URL && typeof target === 'object' && target instanceof URL) {
+      url = target.href
+    }
+    addUrl(this, url)
+
```

Comment:
Missing a check for Data URL:
https://github.com/newrelic/newrelic-browser-agent/blob/e0addb7c19e52429cc17b50a4e6f01892b3c1fc1/feature/xhr/instrument/index.js#L268-L269

Reply:
https://github.com/newrelic/newrelic-browser-agent/projects/7#card-84572817